integration-libs/opf/cta/core/services/opf-dynamic-cta.service.ts

```diff
@@ -0,0 +1,251 @@
+          scriptLocations: [scriptLocation],
+          ctaProductItems: cart.entries?.map((entry) => {
+            return {
+              productId: entry.product?.code,
+              quantity: entry.quantity,
+            } as CtaProductItem;
+          }),
+          additionalData,
+        } as CtaScriptsRequest;
+      })
+    );
+  }
+
+  fillCtaRequestforProductPage(
+    scriptLocation: CtaScriptsLocation,
+    paymentAccountIds: number[]
+  ) {
+    this.isCartPage = false;
+    return combineLatest({
+      product: this.currentProductService.getProduct(),
+      additionalData: this.fillAdditionalData(),
+    }).pipe(
+      take(1),
+      map(({ product, additionalData }) => {
+        return {
+          ctaProductItems: [{ productId: product?.code, quantity: 1 }],
+          paymentAccountIds: paymentAccountIds,
+          scriptLocations: [scriptLocation],
+          additionalData,
+        } as CtaScriptsRequest;
+      })
+    );
+  }
+
+  protected fillAdditionalData(): Observable<Array<KeyValuePair>> {
+    return combineLatest({
+      language: this.languageService.getActive(),
+      currency: this.currencyService.getActive(),
+    }).pipe(
+      take(1),
+      map(({ language, currency }) => {
+        return [
+          {
+            key: 'locale',
+            value: language,
+          },
+          {
+            key: 'currency',
+            value: currency,
+          },
+          {
+            key: 'scriptIdentifier',
+            value: this.getNewScriptIdentifier(),
+          },
+        ];
+      })
+    );
+  }
+
+  initiateEvents() {
+    if (!this.isOnsiteMessagingInit) {
+      this.listenScriptReadyEvent();
+
+      this.isCartPage && this.cartChangedListener();
+      !this.isCartPage && this.productChangedListener();
+
+      this.isOnsiteMessagingInit = true;
+    }
+  }
+
+  stopEvents() {
+    if (this.isOnsiteMessagingInit) {
+      this.subList.forEach((sub) => sub.unsubscribe());
+      this.subList = [];
+      this.globalFunctionsFacade.removeGlobalFunctions(
+        GlobalFunctionsDomain.GLOBAL
+      );
+      this.isOnsiteMessagingInit = false;
+    }
+  }
+
+  protected getNewScriptIdentifier(): string {
+    this.scriptIdentifiers.push(String(this.scriptIdentifiers.length + 1));
+    return String(
+      this.scriptIdentifiers[this.scriptIdentifiers.length - 1]
+    ).padStart(4, '0');
+  }
+
+  registerScriptReadyEvent() {
+    this.globalFunctionsFacade.registerGlobalFunctions({
+      paymentSessionId: '',
+      domain: GlobalFunctionsDomain.GLOBAL,
+    });
+  }
+
+  protected listenScriptReadyEvent() {
+    const sub = this.opfCtaFacade.listenScriptReadyEvent().subscribe({
+      next: (id: string) => {
+        this.isCartPage && this.dispatchCartEvents([id]);
+        !this.isCartPage && this.dispatchProductEvents([id]);
+      },
+    });
+    this.subList.push(sub);
+  }
+
+  protected dispatchCartEvents(scriptIdentifiers: Array<string>): void {
```

Comment:
[MEDIUM]
1. The method names are ambigious to me - whether it dispatches Cart events (e.g. Spartacus events about cart item added etc) OR OPF/CTA events to communicate with global functions.
I believe the latter.
What about adding `Cta` to the name of the method - `dispatchCartEvent` -> `dispatchCtaCartEvent`

Note: analogical remark applies also to the method `dispatchProductEvents` -> `dispatchCtaProductEvents`